### DISCUSSION TITLE
Inconsistent mapping of volume PCA clusters to UMAP plot between `analyze_landscape` and `analyze_landscape_full`

### DISCUSSION THREAD

**Discussion by cbeck22** [zhonge/cryodrgn#415] [General]
Hi!

I am currently using cryoDRGN v3.3.3 and have been working with the analyze_landscape command. From my understanding, this command performs PCA on the 1000 volumes from the kmeans1000 directory and plots each particle cluster onto the UMAP plot generated by the analyze command in the filtering notebook.

After filtering out one outlier, I reran the command on the remaining 999 volumes, which produced the following histograms and plots. As shown, the clusters from the volume PCA align well with the clusters in the UMAP plot:
![image](https://github.com/user-attachments/assets/c2ddf94d-e287-46e4-961a-ee57d14a1068)

However, when I applied the same clustering analysis in the Jupyter notebook generated by the analyze_landscape_full command, I observed a completely different mapping of the same 999 volumes onto the UMAP plot. In this case, the clusters identified by the PCA appear to be randomly dispersed:
![image](https://github.com/user-attachments/assets/a0acbd27-bbfe-4079-8d5c-9fc90cec0a9d)
![image](https://github.com/user-attachments/assets/dd436154-7e6a-4e54-8962-bd388aa4459b)
![image](https://github.com/user-attachments/assets/47b33edf-2b60-46da-b348-5f8f6c6fbfbc)
![image](https://github.com/user-attachments/assets/d4d2949f-cd40-4dc7-8a97-d19cc0145ad1)

I am currently attempting to write my own code in the analyze_landscape_full notebook to plot the clusters onto the UMAP plot manually. Any insights into why the behavior of this notebook differs from the analyze_landscape command would be greatly appreciated!

I'm also not very clear on what exactly `analyze_landscape_full` does. The `state_particle_counts.png` histogram generated by the `analyze_landscape` command already seems to cluster the entire dataset of ~1 million particles. Why does the `analyze_landcape_full` command generate additional training volumes, and why does it use a neural net to assign all of the particles to clusters?

Best,
cbeck

**Comment by cbeck22** [comment]
Actually, I suspect I must have made a mistake somewhere...I'll update this when I find out what I did wrong.

Update:
Ah, I think I understand now. When running `analyze_landscape`, I used the `--vol-ind` option to provide an index .pkl file to filter out an outlier I identified in a previous `analyze_landscape` run. The PCA was then performed on 999 volumes, not the original 1000.

However, `analyze_landscape_full` doesn't have the `--vol-ind` option to filter volumes, and if I manipulate cells 4, 6, and 7:
`centers_ind = np.loadtxt(f'{landscape_dir}/kmeans{K}/centers_ind.txt').astype(int)`
`vol_pc = utils.load_pkl(f'{landscape_dir}/vol_pca_{K}.pkl')`
`kmeans_labels = utils.load_pkl(f'{landscape_dir}/kmeans{K}/labels.pkl')`

to refer to the 999 volumes I filtered using `cryodrgn_utils` and `analyze_landscape`, this results in a discrepancy that messes up the plotting, because some of the particles in the full dataset would have been assigned to the 1 volume that I filtered out.

This made me realize I had some misconceptions about what `analyze_landscape` and `analyze_landscape_full` both do.
1. The clustering of the full dataset in `analyze_landscape` occurs when the kmeans algorithm identifies 1000 clusters. `analyze_landscape` selects a representative volume from each of the 1000 kmeans clusters and runs a PCA directly on the `.mrc` files. Am I right in thinking that the agglomerative clustering algorithm clusters the 1000 representative volumes directly on their `.mrc` files and NOT on the PCA coordinates?

2. I assumed that `analyze_landscape_full` uses agglomerative clustering to cluster the full dataset of particles after doing the PCA analysis on the whole dataset. In reality, the Jupyter notebook just fetches the kmeans1000 clusters calculated by `analyze_landscape` and plots those.

So, a few questions:
1. Is there a reason why the notebook doesn't run agglomerative clustering on the full dataset? Is it too computationally intensive?
2. What is the purpose of the neural net for mapping the full dataset to the PCA landscape calculated by `analyze_landscape` for the 1000 kmeans representative volumes? Is it too computationally intensive to run PCA on the `.mrc` files of the full dataset, and does the neural net somehow interpolate what PCA values the rest of the dataset's particles should have?
3. Would you be able to add some more information to the `analyze_landscape_full` documentation to provide some more information on what it does?

**Comment by michal-g** [comment]
Hey, I'm making this a Discussion as there isn't an outright issue being described here, although as I've mentioned in e.g. #413 we are working on some overhauls of the `analysis_landscape` commands as well as an updated document that should address your point re: `analyze_landscape_full` documentation.

I will get back to soon about your other points, but in the meantime, it would be helpful if you took a look at some of our older material for the landscape analyses:
[gitbook tutorial on landscape analysis](https://ez-lab.gitbook.io/cryodrgn/cryodrgn-conformational-landscape-analysis)
[cryoDRGN_landscape_analysis_thesis_chapter.pdf](https://ezlab.cs.princeton.edu/assets/pdf/zhong-phd-csb-2022.pdf)

**Comment by zhonge** [comment]
Very cool results, and glad you found the bug about the vol index filtering in the jupyter notebook! That's a good catch, we could add in a cell in the jupyter notebook that asks for a volume filtering.

As for `cryodrgn analyze_landscape_full`, that step is learning a volume-space embedding for **all particles** in the dataset. To briefly summarize, the first `analyze_landscape` step generates and analyzes the 1000 .mrc files. Specifically, agglomerating clustering is performed directly on the volumes to infer discrete states; PCA is performed on the 1000 volumes to infer the top 5 (by default) principal components of the set of volumes. The analyze_landscape_full step extends the PCA analysis by populating the volume PCA plots from the initial 1,000 points to encompass the entire dataset.
 
To answer your first two questions:
1. Yes, agglomerative clustering is done directly on the 1000 volumes, not on the latent or PCA representations. 
2. `analyze_landscape_full` is not related to the agglomerative clustering (the discrete interpretation). It is filling out the continuous representation of the dataset in volume PCA space.

To answer your next three questions:
1. It is too expensive to run agglomerative clustering on all the volumes. The point of generating 1000 volumes is to generate a representative set of volumes that can be tractably analyzed. A key assumption is each volume from the kmeans1000 is representative of the underlying particles that were clustered.
2. Yes and yes
3. The documentation for the `analyze_landscape_full` is sparse and we can fill it in!

**Comment by cbeck22** [comment]
Thank you both so much for your replies! I should've remembered to check that landscape analysis thesis chapter first, it was very informative and answered all of my questions.

Michal, since you mentioned that you're working on overhauling the landscape analysis commands and notebook, could I make a few suggestions as a user?
1. It would be fantastic if the `analyze_landscape` command came with an analysis notebook similar to`analyze_landscape_full`. In particular, it would be helpful if there was a way to display all of the `.png` files for the annotated volume PCA and UMAP analyses side-by-side, making it easier to compare different clustering parameters. For example, I wrote a function that takes the path to a clustering directory as input and displays the `.pngs`. The function calls look like:
`display_analyze_landscape('landscape_sketch-2/clustering_L2_ward_2/')`
`display_analyze_landscape('landscape_sketch-2/clustering_L2_ward_3/')`,
And it displays the `.pngs` like this:
![image](https://github.com/user-attachments/assets/730873e9-405f-4034-9035-fe40ad092743)

2. When I first used `analyze_landscape`, it wasn't clear how to organize the outputs from different `analyze_landscape` calls and make sure I wasn't overwriting files from a previous command. It could be useful if the documentation included a suggested file hierarchy? The following file structure worked for me:
- **landscape.99**
  - **landscape_sketch-1/**: Contains all the files output by an `analyze_landscape` command. Specified using the `-o` flag for the output directory
    - **kmeans1000**/: Created with the first `analyze_landscape` call 
    - **clustering_L2_average_10/**: Default clustering analysis created with the first `analyze_landscape` call
    - **clustering_L2_ward_2/**: Second clustering analysis created with a second `analyze_landscape` call with `--skip-vols` enabled
    - **landscape_full-1/**: Contains the files output by an `analyze_landscape_full` command. Specified using the `-o` flag.
    - mask.mrc, .png files, .pkl files, etc
  - **landscape_sketch-2/**: Contains the files for another `analyze_landscape` command with `--skip-vols` enabled, but maybe with a different mask or particle filtering
    - **kmeans1000**/: Symlink to the same directory in `landscape_sketch-2`. I found that the `-o` directory needs to have a `kmeans` directory in it for `analyze_landscape` to run, which isn't clear from the documentation

3. Another small suggestion: could `analyze_landscape` automatically use the same pixel size used by the `train_vae` command? It would make it a little easier to align atomic models to the maps or create custom masks.

4. In the `analyze_landscape_full` notebook, I had difficulty comparing how clusters looked across 1) the latent embeddings PCA, 2) latent embeddings UMAP, and 3) volume PCA. For example, the cluster coloring scheme isn't standardized between the cluster histograms and how they're plotted on the volume PCA or the latent embeddings UMAP. In addition, it became difficult comparing the results of different clustering analyses because it involved a lot of copying and pasting the code to make the scatterplots and manually changing several parameters in each instance. Would you consider implementing a plotting function that takes a clustering assignment as input and automatically makes the three plots using a standardized coloring scheme?

To implement this myself, I created a ClusteringAnalysis object that's initialized using the path to a `landscape_sketch` directory and automatically pulls information from the directory like the latent embeddings, PCA, UMAP, and volume PCA and compiles them into a DataFrame:
![image](https://github.com/user-attachments/assets/cd3811c6-9ae8-4d55-beab-e509cef2974e)

I then use this to attach clustering assignments for each particle to the columns at the end of the DataFrame, which makes it simple to auto-generate plots colored by cluster:
![image](https://github.com/user-attachments/assets/17054e4f-d7cc-46f0-bfc0-bcf07cdc4bc0)
![image](https://github.com/user-attachments/assets/d6671eea-5673-41e2-9209-16a9bf3dedf9)
![image](https://github.com/user-attachments/assets/0826d9c6-51cb-47b2-aced-1dfb9fe1ba11)

Having tools like these built-in to the notebook would really help with analyzing the data that cryoDRGN outputs. In particular, having the clustering assignments stored as a DataFrame makes it really easy to load the clustering assignments using cryoSPARC's `cryosparc-tools` Jupyter notebook, which I can use to automate importing groups of particles into a cryoSPARC project. This gets around the need to export a `.cs` file from cryoSPARC and manually edit `.csg` files for potentially dozens of clusters of particles, as described in #150 :
![image](https://github.com/user-attachments/assets/6eaafd07-f7a1-43cc-8ebf-b13ffad7d975)
![image](https://github.com/user-attachments/assets/d404f121-e9f3-4806-a542-3d7bceb0038a)

Thanks you so much for all your hard work on cryoDRGN! It's been extremely valuable for my project, and I really appreciate the work you've put into maintaining it. 

Cheers,
cbeck
